### PR TITLE
Control Allocation Sequential Desaturation unit tests

### PR DESCRIPTION
**Pull request by somebody-once-told-me**
### Solved Problem
Unit tests for ControlAllocationSequentialDesaturation. No code changes are made to any existing library, this only adds unit tests. The unit tests serve 2 purposes: 1) help ensure that expected functionality remains the same and 2) documents and demonstrates (through examples) the desaturation algorithm.

### CODE REVIEW

**Comment by somebody-once-told-me** [comment]
hmm, looks like the tailsitter SITL test is failing? curious as this only adds a unit test.

**Comment by somebody-once-told-me** [comment]
merged; now all SITL tests are passing.

**Comment by sfuhrer** [comment]
Thanks for your contribution, I'm going through it later.

>hmm, looks like the tailsitter SITL test is failing? curious as this only adds a unit test.

It's now passing. These SITL tests are a bit flaky. 

Can you rebase your branch on latest main? Something seems to have been messed up with your history, and this PR shows up many commits that don't come from you.

**Comment by somebody-once-told-me** [comment]
@sfuhrer I can try, but I am not sure why those other commits are showing up. Although, the changes are accurately captured, as can be seen in the `Files changed` tab. If this PR is merged as-is, won't it all be squashed in a single commit anyway (and the weird commit history shown above won't matter)?

**Comment by somebody-once-told-me** [comment]
@sfuhrer the commit history has been cleaned up.

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:163]:
```suggestion
	constexpr float THRUST_Z_TOTAL{-0.75f};
```

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:176]:
```suggestion
	constexpr float THRUST_Z_PER_MOTOR{-THRUST_Z_TOTAL / 4.f};
```
Generally best practice to only use floats in divisions.

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:212]: Could you maybe define the "5" as a constant by itself?

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:329]: @MaEtUgR maybe it's a good time to document this magic number in the algorithm as well!

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [reply src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:163]: I generally like being very explicit.

**Review by sfuhrer — COMMENTED** [review]
Thanks for kicking this off, that's very valuable. 
I only have some minor things. I leave it up to you to also add some test cases for the "normalized" allocation - could though also be a follow up PR.

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (somebody-once-told-me) [reply src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:122]: good question; after looking into it, looks like `EXPECT_FLOAT_EQ` applies tolerance of 1e-4, so I will use that.

Comment (somebody-once-told-me) [reply src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:122]: apparently `EXPECT_FLOAT_EQ` applies a tighter tolerance. I was getting a failure for values 0.30000025 and 0.30000004. So, I am using `EXPECT_NEAR` with tolerance 1e-4.

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (somebody-once-told-me) [reply src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:163]: is this something that can be flagged in `make format`?

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:121]:
```suggestion
static constexpr float EXPECT_NEAR_TOL{1e-4f};
```

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:132]:
```suggestion
	control_sp(ControlAllocation::ControlAxis::ROLL) = 0.f;
```

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:170]:
```suggestion
	constexpr int MOTOR_COUNT{4};
```
As the motor count is a countable physical object. I know I previously advised to use /4.f, but if you pull it out as a variable like you do now (which is a good thing!) then I would not cover a countable physical object by a float but rather have float/integer in the math.

**Review by sfuhrer — COMMENTED** [review]
Just some mini comments, otherwise good to go.

PS: please do not force push a branch that you opened already a pull request for, as otherwise it's not possible for the reviewer to see what has changed since the last review. Just add the review comments as new commits on top, and push the branch like that. Once the PR is approved you can clean up the commits or we just squash-merge to merge it all into one commit.

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:210]: Same as above

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:251]: Same as above

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:288]: Same as above

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:312]: Same as above

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (sfuhrer) [inline src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:368]: Same as above

**Review by sfuhrer — COMMENTED** [review]
You forgot a couple of MOTOR_COUNTS. Lets do them all in ints and then we can get this in!

**Comment by somebody-once-told-me** [comment]
@sfuhrer thanks, I fixed those and also added the float specifier suffix to all float zeros, replaced magic numbers with `MOTOR_COUNT` where applicable, and changed `YAW_MOTORS` to an int.

**Review by sfuhrer — APPROVED** [review]
Thanks for the last clean up, let's get it in!

File: src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp
Comment (MaEtUgR) [reply src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:329]: I can only imagine why this is useful. I think @bresch had the full context here.

**Pull request merged**
